yarn-project/circuits.js/src/structs/rollup/root_rollup.ts

```diff
@@ -131,18 +85,33 @@ export class RootRollupInputs {
  */
 export class RootRollupPublicInputs {
   constructor(
-    /** Snapshot of archive tree after this block/rollup been processed */
-    public archive: AppendOnlyTreeSnapshot,
-    /** The root for the protocol circuits vk tree */
+    /** Snapshot of archive tree before/after this rollup been processed */
+    public previousArchive: AppendOnlyTreeSnapshot,
+    public endArchive: AppendOnlyTreeSnapshot,
+    public previousBlockHash: Fr,
+    public endBlockHash: Fr,
+    // This is a u64 in nr, but GlobalVariables contains this as a u64 and is mapped to ts as a field, so I'm doing the same here
+    public endTimestamp: Fr,
```

Comment:
Hmm where are we validating that the timestamp of the first block in the epoch is greater than the timestamp of the last block in the previous epoch? Feels like I may have missed that in the design.

Reply:
Should we be rehydrating that previous header using the `previousBlockhash` inside the circuit to make that check?

Reply:
Actually, should we be rehydrating that header to make the `assert_prev_block_rollups_follow_on_from_each_other` check to ensure both epochs "glue" together correctly?

Reply:
Yep correct - we are not currently checking the timestamp of last block in prev epoch vs first block in new epoch. Imho `assert_prev_block_rollups_follow_on_from_each_other` wouldn't be the best place as there we are checking two groups of blocks that will end up in a single epoch together. Checking the 'first' timestamp would be wasted gates most of the time.
I think we could either:

- As you suggest, recreate the previous block's header using `previousBlockhash` inside the final `root` circuit and check there, or
- Check it on L1

I don't fully understand the new `getTimestampForSlot` code in `Rollup.sol`, but using that we can gather the start timestamp of the current epoch and the end timestamp of the previous without adding extra PIs.

Reply:
> As you suggest, recreate the previous block's header using previousBlockhash inside the final root circuit and check there, or

I'd go with this one in order to save L1 gas. But we can push this for another PR, so this one doesn't keep growing.